File: src/cursor_orch/state.py
from __future__ import annotations

import json
from datetime import datetime, timezone
from dataclasses import asdict, dataclass, field
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from cursor_orch.api.repo_store import RepoStoreClient
    from cursor_orch.config import OrchestratorConfig

MAX_EVENTS_BYTES = 256 * 1024
ROTATE_KEEP_BYTES = 128 * 1024


@dataclass
class AgentState:
    task_id: str
    agent_id: str | None = None
    status: str = "pending"
    started_at: str | None = None
    finished_at: str | None = None
    pr_url: str | None = None
    summary: str | None = None
    blocked_reason: str | None = None
    blocked_since: str | None = None
    retry_count: int = 0


@dataclass
class LifecycleAgentState:
    node_id: str
    label: str
    kind: str
    status: str = "pending"
    parent_node_id: str | None = None
    task_id: str | None = None
    agent_id: str | None = None
    started_at: str | None = None
    finished_at: str | None = None


@dataclass
class OrchestrationState:
    orchestration_id: str
    run_id: str
    orchestrator_agent_id: str | None = None
    status: str = "pending"
    started_at: str | None = None
    agents: dict[str, AgentState] = field(default_factory=dict)
    main_agent: LifecycleAgentState | None = None
    phase_agents: dict[str, LifecycleAgentState] = field(default_factory=dict)
    task_phase_map: dict[str, str] = field(default_factory=dict)
    error: str | None = None


@dataclass
class OrchestrationEvent:
    timestamp: str
    event_type: str
    task_id: str | None = None
    phase_id: str | None = None
    agent_node_id: str | None = None
    agent_kind: str | None = None
    detail: str = ""
    payload: dict[str, str] = field(default_factory=dict)


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


def _default_phase_agents() -> dict[str, LifecycleAgentState]:
    return {
        "planning": LifecycleAgentState(
            node_id="phase-planning",
            label="Planning",
            kind="phase",
            parent_node_id="main-orchestrator",
        ),
        "scheduling": LifecycleAgentState(
            node_id="phase-scheduling",
            label="Scheduling",
            kind="phase",
            parent_node_id="main-orchestrator",
        ),
        "execution": LifecycleAgentState(
            node_id="phase-execution",
            label="Execution",
            kind="phase",
            parent_node_id="main-orchestrator",
        ),
        "finalization": LifecycleAgentState(
            node_id="phase-finalization",
            label="Finalization",
            kind="phase",
            parent_node_id="main-orchestrator",
        ),
    }


def create_initial_state(config: OrchestratorConfig, run_id: str) -> OrchestrationState:
    agents = {
        task.id: AgentState(task_id=task.id)
        for task in config.tasks
    }
    state = OrchestrationState(
        orchestration_id=run_id,
        run_id=run_id,
        agents=agents,
    )
    ensure_lifecycle_agents(state)
    return state


def ensure_lifecycle_agents(state: OrchestrationState) -> None:
    if state.main_agent is None:
        state.main_agent = LifecycleAgentState(
            node_id="main-orchestrator",
            label="Main Orchestrator",
            kind="main",
            status=state.status,
            agent_id=state.orchestrator_agent_id,
            started_at=state.started_at,
        )
    else:
        if state.main_agent.agent_id is None and state.orchestrator_agent_id:
            state.main_agent.agent_id = state.orchestrator_agent_id
        if state.main_agent.started_at is None and state.started_at:
            state.main_agent.started_at = state.started_at
    if not state.phase_agents:
        state.phase_agents = _default_phase_agents()
    else:
        defaults = _default_phase_agents()
        for phase_id, default in defaults.items():
            state.phase_agents.setdefault(phase_id, default)


def seed_main_agent(
    state: OrchestrationState,
    *,
    agent_id: str | None,
    status: str,
    started_at: str | None = None,
) -> None:
    ensure_lifecycle_agents(state)
    state.orchestrator_agent_id = agent_id
    if state.main_agent is None:
        return
    state.main_agent.agent_id = agent_id
    state.main_agent.status = status
    if started_at:
        state.main_agent.started_at = started_at
    elif state.main_agent.started_at is None:
        state.main_agent.started_at = _now_iso()


def set_phase_status(
    state: OrchestrationState,
    phase_id: str,
    status: str,
    *,
    timestamp: str | None = None,
) -> None:
    ensure_lifecycle_agents(state)
    phase = state.phase_agents.get(phase_id)
    if phase is None:
        phase = LifecycleAgentState(
            node_id=f"phase-{phase_id}",
            label=phase_id.replace("-", " ").title(),
            kind="phase",
            parent_node_id="main-orchestrator",
        )
        state.phase_agents[phase_id] = phase
    phase.status = status
    if status in ("running", "launching") and phase.started_at is None:
        phase.started_at = timestamp or _now_iso()
    if status in ("finished", "failed", "stopped"):
        phase.finished_at = timestamp or _now_iso()


def assign_task_phase(state: OrchestrationState, task_id: str, phase_id: str) -> None:
    ensure_lifecycle_agents(state)
    state.task_phase_map[task_id] = phase_id


def serialize(state: OrchestrationState) -> str:
    return json.dumps(asdict(state), indent=2)


def _lifecycle_agent_from_dict(d: dict, default_node_id: str, default_label: str, default_kind: str) -> LifecycleAgentState:
    return LifecycleAgentState(
        node_id=d.get("node_id") or default_node_id,
        label=d.get("label") or default_label,
        kind=d.get("kind") or default_kind,
        status=d.get("status", "pending"),
        parent_node_id=d.get("parent_node_id"),
        task_id=d.get("task_id"),
        agent_id=d.get("agent_id"),
        started_at=d.get("started_at"),
        finished_at=d.get("finished_at"),
    )


def deserialize(json_str: str) -> OrchestrationState:
    raw = json.loads(json_str)
    if "gist_id" in raw and "run_id" not in raw:
        raw["run_id"] = raw.pop("gist_id")
    elif "gist_id" in raw:
        raw.pop("gist_id")
    agents = {
        k: AgentState(**v)
        for k, v in raw.pop("agents", {}).items()
    }
    raw_main_agent = raw.pop("main_agent", None)
    main_agent = (
        _lifecycle_agent_from_dict(raw_main_agent, "main-orchestrator", "Main Orchestrator", "main")
        if isinstance(raw_main_agent, dict) and raw_main_agent
        else None
    )
    phase_agents = {}
    for phase_id, phase_data in raw.pop("phase_agents", {}).items():
        if not isinstance(phase_data, dict):
            continue
        default_node_id = f"phase-{phase_id}"
        default_label = phase_id.replace("-", " ").title()
        phase_agents[phase_id] = _lifecycle_agent_from_dict(phase_data, default_node_id, default_label, "phase")
    task_phase_map = {
        str(task_id): str(phase_id)
        for task_id, phase_id in raw.pop("task_phase_map", {}).items()
    }
    state = OrchestrationState(
        **raw,
        agents=agents,
        main_agent=main_agent,
        phase_agents=phase_agents,
        task_phase_map=task_phase_map,
    )
    ensure_lifecycle_agents(state)
    return state


def serialize_event(event: OrchestrationEvent) -> str:
    return json.dumps(asdict(event), separators=(",", ":"))


def deserialize_event(json_str: str) -> OrchestrationEvent:
    return OrchestrationEvent(**json.loads(json_str))


def sync_to_repo(repo_store: RepoStoreClient, run_id: str, state: OrchestrationState) -> None:
    repo_store.write_file(run_id, "state.json", serialize(state))


def sync_from_repo(repo_store: RepoStoreClient, run_id: str) -> OrchestrationState:
    content = repo_store.read_file(run_id, "state.json")
    return deserialize(content)


def append_event(repo_store: RepoStoreClient, run_id: str, event: OrchestrationEvent) -> None:
    existing = repo_store.read_file(run_id, "events.jsonl")
    line = serialize_event(event)
    content = f"{existing}{line}\n" if existing else f"{line}\n"
    content = _rotate_events(content)
    repo_store.write_file(run_id, "events.jsonl", content)


def _rotate_events(content: str) -> str:
    if len(content.encode("utf-8")) <= MAX_EVENTS_BYTES:
        return content
    encoded = content.encode("utf-8")
    tail = encoded[-ROTATE_KEEP_BYTES:]
    text = tail.decode("utf-8", errors="replace")
    first_newline = text.find("\n")
    if first_newline >= 0:
        return text[first_newline + 1:]
    return text


def read_events(repo_store: RepoStoreClient, run_id: str) -> list[OrchestrationEvent]:
    content = repo_store.read_file(run_id, "events.jsonl")
    if not content.strip():
        return []
    events: list[OrchestrationEvent] = []
    for line in content.strip().split("\n"):
        stripped = line.strip()
        if not stripped:
            continue
        try:
            events.append(deserialize_event(stripped))
        except (json.JSONDecodeError, TypeError):
            continue
    return events

File: src/state.ts
import type { RepoStoreClient } from "./api/repo-store.js";
import type { OrchestratorConfig } from "./config/types.js";

export const MAX_EVENTS_BYTES = 256 * 1024;
export const ROTATE_KEEP_BYTES = 128 * 1024;

export interface AgentState {
  task_id: string;
  agent_id: string | null;
  status: string;
  started_at: string | null;
  finished_at: string | null;
  pr_url: string | null;
  summary: string | null;
  blocked_reason: string | null;
  blocked_since: string | null;
  retry_count: number;
}

export interface LifecycleAgentState {
  node_id: string;
  label: string;
  kind: string;
  status: string;
  parent_node_id: string | null;
  task_id: string | null;
  agent_id: string | null;
  started_at: string | null;
  finished_at: string | null;
}

export interface OrchestrationState {
  orchestration_id: string;
  run_id: string;
  orchestrator_agent_id: string | null;
  status: string;
  started_at: string | null;
  delegation_phase_index: number | null;
  agents: Record<string, AgentState>;
  main_agent: LifecycleAgentState | null;
  phase_agents: Record<string, LifecycleAgentState>;
  task_phase_map: Record<string, string>;
  error: string | null;
}

export interface OrchestrationEvent {
  timestamp: string;
  event_type: string;
  task_id: string | null;
  phase_id: string | null;
  agent_node_id: string | null;
  agent_kind: string | null;
  detail: string;
  payload: Record<string, string>;
}

function nowIso(): string {
  return new Date().toISOString();
}

function defaultPhaseAgents(): Record<string, LifecycleAgentState> {
  return {
    planning: {
      node_id: "phase-planning",
      label: "Planning",
      kind: "phase",
      status: "pending",
      parent_node_id: "main-orchestrator",
      task_id: null,
      agent_id: null,
      started_at: null,
      finished_at: null,
    },
    scheduling: {
      node_id: "phase-scheduling",
      label: "Scheduling",
      kind: "phase",
      status: "pending",
      parent_node_id: "main-orchestrator",
      task_id: null,
      agent_id: null,
      started_at: null,
      finished_at: null,
    },
    execution: {
      node_id: "phase-execution",
      label: "Execution",
      kind: "phase",
      status: "pending",
      parent_node_id: "main-orchestrator",
      task_id: null,
      agent_id: null,
      started_at: null,
      finished_at: null,
    },
    finalization: {
      node_id: "phase-finalization",
      label: "Finalization",
      kind: "phase",
      status: "pending",
      parent_node_id: "main-orchestrator",
      task_id: null,
      agent_id: null,
      started_at: null,
      finished_at: null,
    },
  };
}

export function createInitialState(config: OrchestratorConfig, runId: string): OrchestrationState {
  const agents: Record<string, AgentState> = {};
  for (const task of config.tasks) {
    agents[task.id] = { task_id: task.id, agent_id: null, status: "pending", started_at: null, finished_at: null, pr_url: null, summary: null, blocked_reason: null, blocked_since: null, retry_count: 0 };
  }
  const state: OrchestrationState = {
    orchestration_id: runId,
    run_id: runId,
    orchestrator_agent_id: null,
    status: "pending",
    started_at: null,
    delegation_phase_index: null,
    agents,
    main_agent: null,
    phase_agents: {},
    task_phase_map: {},
    error: null,
  };
  ensureLifecycleAgents(state);
  return state;
}

export function ensureLifecycleAgents(state: OrchestrationState): void {
  if (state.main_agent === null) {
    state.main_agent = {
      node_id: "main-orchestrator",
      label: "Main Orchestrator",
      kind: "main",
      status: state.status,
      parent_node_id: null,
      task_id: null,
      agent_id: state.orchestrator_agent_id,
      started_at: state.started_at,
      finished_at: null,
    };
  } else {
    if (state.main_agent.agent_id === null && state.orchestrator_agent_id) {
      state.main_agent.agent_id = state.orchestrator_agent_id;
    }
    if (state.main_agent.started_at === null && state.started_at) {
      state.main_agent.started_at = state.started_at;
    }
  }
  if (Object.keys(state.phase_agents).length === 0) {
    state.phase_agents = defaultPhaseAgents();
  } else {
    const defaults = defaultPhaseAgents();
    for (const [phaseId, def] of Object.entries(defaults)) {
      if (!state.phase_agents[phaseId]) {
        state.phase_agents[phaseId] = def;
      }
    }
  }
}

export function seedMainAgent(
  state: OrchestrationState,
  opts: { agent_id: string | null; status: string; started_at?: string | null },
): void {
  ensureLifecycleAgents(state);
  state.orchestrator_agent_id = opts.agent_id;
  if (state.main_agent === null) return;
  state.main_agent.agent_id = opts.agent_id;
  state.main_agent.status = opts.status;
  if (opts.started_at) {
    state.main_agent.started_at = opts.started_at;
  } else if (state.main_agent.started_at === null) {
    state.main_agent.started_at = nowIso();
  }
}

export function setPhaseStatus(
  state: OrchestrationState,
  phaseId: string,
  status: string,
  opts?: { timestamp?: string | null },
): void {
  ensureLifecycleAgents(state);
  let phase = state.phase_agents[phaseId];
  if (!phase) {
    phase = {
      node_id: `phase-${phaseId}`,
      label: phaseId.replace(/-/g, " ").replace(/\b\w/g, (c) => c.toUpperCase()),
      kind: "phase",
      status: "pending",
      parent_node_id: "main-orchestrator",
      task_id: null,
      agent_id: null,
      started_at: null,
      finished_at: null,
    };
    state.phase_agents[phaseId] = phase;
  }
  phase.status = status;
  const ts = opts?.timestamp ?? null;
  if ((status === "running" || status === "launching") && phase.started_at === null) {
    phase.started_at = ts ?? nowIso();
  }
  if (status === "finished" || status === "failed" || status === "stopped") {
    phase.finished_at = ts ?? nowIso();
  }
}

export function assignTaskPhase(state: OrchestrationState, taskId: string, phaseId: string): void {
  ensureLifecycleAgents(state);
  state.task_phase_map[taskId] = phaseId;
}

export function serialize(state: OrchestrationState): string {
  return JSON.stringify(state, null, 2);
}

function lifecycleFromDict(d: Record<string, unknown>, defaultNodeId: string, defaultLabel: string, defaultKind: string): LifecycleAgentState {
  return {
    node_id: (d.node_id as string) || defaultNodeId,
    label: (d.label as string) || defaultLabel,
    kind: (d.kind as string) || defaultKind,
    status: (d.status as string) || "pending",
    parent_node_id: (d.parent_node_id as string) ?? null,
    task_id: (d.task_id as string) ?? null,
    agent_id: (d.agent_id as string) ?? null,
    started_at: (d.started_at as string) ?? null,
    finished_at: (d.finished_at as string) ?? null,
  };
}

export function deserialize(jsonStr: string): OrchestrationState {
  const raw = JSON.parse(jsonStr) as Record<string, unknown>;
  if ("gist_id" in raw && !("run_id" in raw)) {
    raw.run_id = raw.gist_id;
    delete raw.gist_id;
  } else if ("gist_id" in raw) {
    delete raw.gist_id;
  }
  const agentsRaw = (raw.agents as Record<string, Record<string, unknown>>) || {};
  const agents: Record<string, AgentState> = {};
  for (const [k, v] of Object.entries(agentsRaw)) {
    agents[k] = v as unknown as AgentState;
  }
  const rawMain = raw.main_agent as Record<string, unknown> | null | undefined;
  const mainAgent =
    rawMain && typeof rawMain === "object"
      ? lifecycleFromDict(rawMain, "main-orchestrator", "Main Orchestrator", "main")
      : null;
  const phaseAgents: Record<string, LifecycleAgentState> = {};
  const phaseRaw = (raw.phase_agents as Record<string, Record<string, unknown>>) || {};
  for (const [phaseId, phaseData] of Object.entries(phaseRaw)) {
    if (typeof phaseData !== "object" || !phaseData) continue;
    const defaultNodeId = `phase-${phaseId}`;
    const defaultLabel = phaseId.replace(/-/g, " ").replace(/\b\w/g, (c) => c.toUpperCase());
    phaseAgents[phaseId] = lifecycleFromDict(phaseData, defaultNodeId, defaultLabel, "phase");
  }
  const taskPhaseMap: Record<string, string> = {};
  const tpm = (raw.task_phase_map as Record<string, string>) || {};
  for (const [a, b] of Object.entries(tpm)) {
    taskPhaseMap[String(a)] = String(b);
  }
  const state: OrchestrationState = {
    orchestration_id: String(raw.orchestration_id ?? ""),
    run_id: String(raw.run_id ?? ""),
    orchestrator_agent_id: (raw.orchestrator_agent_id as string) ?? null,
    status: String(raw.status ?? "pending"),
    started_at: (raw.started_at as string) ?? null,
    delegation_phase_index: typeof raw.delegation_phase_index === "number" ? raw.delegation_phase_index : null,
    agents,
    main_agent: mainAgent,
    phase_agents: phaseAgents,
    task_phase_map: taskPhaseMap,
    error: (raw.error as string) ?? null,
  };
  ensureLifecycleAgents(state);
  return state;
}

export function serializeEvent(event: OrchestrationEvent): string {
  return JSON.stringify(event);
}

export function deserializeEvent(jsonStr: string): OrchestrationEvent {
  return JSON.parse(jsonStr) as OrchestrationEvent;
}

export async function syncToRepo(repoStore: RepoStoreClient, runId: string, state: OrchestrationState): Promise<void> {
  await repoStore.writeFile(runId, "state.json", serialize(state));
}

export async function syncFromRepo(repoStore: RepoStoreClient, runId: string): Promise<OrchestrationState> {
  const content = await repoStore.readFile(runId, "state.json");
  return deserialize(content);
}

export async function appendEvent(repoStore: RepoStoreClient, runId: string, event: OrchestrationEvent): Promise<void> {
  const line = serializeEvent(event);
  await repoStore.updateFile(runId, "events.jsonl", (current) => {
    const nextContent = current ? `${current}${line}\n` : `${line}\n`;
    return rotateEvents(nextContent);
  });
}

function rotateEvents(content: string): string {
  if (Buffer.byteLength(content, "utf8") <= MAX_EVENTS_BYTES) {
    return content;
  }
  const encoded = Buffer.from(content, "utf8");
  const tail = encoded.subarray(Math.max(0, encoded.length - ROTATE_KEEP_BYTES));
  let text = tail.toString("utf8");
  const firstNl = text.indexOf("\n");
  if (firstNl >= 0) {
    return text.slice(firstNl + 1);
  }
  return text;
}

export async function readEvents(repoStore: RepoStoreClient, runId: string): Promise<OrchestrationEvent[]> {
  const content = await repoStore.readFile(runId, "events.jsonl");
  if (!content.trim()) return [];
  const events: OrchestrationEvent[] = [];
  for (const line of content.trim().split("\n")) {
    const stripped = line.trim();
    if (!stripped) continue;
    try {
      events.push(deserializeEvent(stripped));
    } catch {
      continue;
    }
  }
  return events;
}
